### ISSUE TITLE
Motion controls stopped working after Android System WebView 76 update

### ISSUE THREAD

**Issue by pzmarzly** [rmst/yoke#39]
Fun fact: Yoke was created as no phone-as-PC-gamepad app on the Play Store used to capture device movement.

> onDeviceOrientation events are never fired now

Details: https://github.com/rmst/yoke/pull/38#issuecomment-521857892

If you are affected by this, please try [downgrading ASW](https://www.apkmirror.com/apk/google-inc/android-system-webview/android-system-webview-75-0-3770-101-release/) for now and describe your results here.

From Chrome DevTools (remote device inspection over ADB):

> [Deprecation] The `devicemotion` event is deprecated on insecure origins and will be removed in M76, around July 2019. Event handlers can still be registered but are no longer invoked since M74, around April 2019. See https://www.chromestatus.com/feature/5688035094036480 for more details.

Suggestion from [Google ~announcement~ "deprecation proposal"](https://www.chromium.org/Home/chromium-security/deprecating-powerful-features-on-insecure-origins):

> On a local network, you can test on your Android device using port forwarding to access a remote host as localhost.

... which sucks as phone would usually need to be connected over USB (ADB over Wifi is disabled by most vendor overlays).

Perhaps we could add a proxy server hosted on the phone (since `localhost` is a secure origin) but this seems like a lot of work. It would be a easier to [download the files to the phone](https://stackoverflow.com/a/10666892/5108318) (I hope `file://` is a secure origin, but am I right?).

Apparently we [cannot unconditionally accept unsigned certs, even for LAN](https://stackoverflow.com/q/5977977/5108318), [since 2016](https://stackoverflow.com/a/38389319/5108318), if we want the app to be on the Play Store (and, AFAIK, as it's usually the case with Google, there is no appeal process where we could explain it's LAN connection). Maybe we could add "certificate error" dialog that users would need to click through every time they open the app, just to satisfy Play Store algorithms.

Another option is to once again fetch orientation events in Java code and everything else in JS, but I'd like to avoid that.

Personally, I'd go for downloading files to phone storage, but I'd like to hear your opinions.

**Comment by medape** [comment]
After thinking it over... I agree with you. I haven't checked if `file://` is considered a secure origin yet, but judging by the SO's you linked to, it seems so.

My first instinct was to propose bundling a proxy server with the Yoke app, but besides being too complex and possibly slower, it could potentially get our app pulled from Google Play for circunventing the secure origins policy. (If not now, possibly later if they ever change their algorithms, which they often do.)

Regarding the rest of the methods, forcing users to install ADB on their computer, or enabling it in their phone, or juggling security certificates is needlessly complex, and probably a very bad idea from security's viewpoint.

It's much simpler to ditch the web server and load everything from external storage as you said. It's also the only way each user can customize separately their gamepads. (With the current system, configuration is per computer, and every user is forced to use the same base code and layouts.)

I only have a small question: what should the app do if permission to access external storage isn't granted? Should it quit with an error? Should it use a bundled copy of the gamepad?

**Comment by pzmarzly** [comment]
> I only have a small question: what should the app do if permission to access external storage isn't granted? Should it quit with an error? Should it use a bundled copy of the gamepad?

AFAIK apps don't need permissions to write to `/sdcard/Android/data/<appname>/files` (internal storage, called `sdcard` for legacy reasons) or even `/data/media/<SD-CARD-ID>/Android/data/<appname>/files`. But if I'm wrong, I think app should quit with an error.

**Comment by rmst** [comment]
Wouldn't it make sense to just revert the last change in the Android app that removes the native motion control. In case we don't want to append it to each dataframe as was previously done, we have the option to create a custom webview interface that would allow us to forward the motion sensor data from Java to Javascript (https://developer.android.com/guide/webapps/webview#BindingJavaScript).

**Comment by pzmarzly** [comment]
@rmst putting accel. reading to Java code is a viable option, though maybe not the most elegant (on Python side, which would need to reorder the readings). I'm against sending data from Java to Javascript due to latency.

If we were to revert the latest changes in Java code (or rather just https://github.com/rmst/yoke-android/pull/3/commits/d87233de3455ddd63c6cfecfacf7261043224faa as https://github.com/rmst/yoke-android/pull/3 had some unrelated changes too), we would need to cut the float accuracy there (so whole floats wouldn't be sent (11 digits or something), but only the first few digits) or, better yet, remap these floats to integers like JS code [does now](https://github.com/rmst/yoke/blob/b887848b905a4d30ab65445881e8ce1bd2b05efa/yoke/assets/joypad/base.js#L363).

@medape what do you think?

**Comment by rmst** [comment]
I wasn't aware that WebAppInterface introduces latency. Thought we were using the same approach to transmit the other values back to the yoke server via UDP. Agree, that reverting would be to extreme. Just reintroducing the few lines that did the native motion sensing would be enough.

**Comment by pzmarzly** [comment]
> I wasn't aware that WebAppInterface introduces noticeable latency.

Noticeable - probably not. But I expect there is >1ms of it. Maybe I'm wrong about that. I think so because FFI boundaries generally induce some latency (even more so between 2 GC languages), and because I expect some JS shenanigans (e.g. global string variable preventing some JIT optimizations, event loop ticks being postponed, etc.). I can't find any benchmarks on the web, though.

> we were using the same approach to transmit the other values back to the yoke server via UDP

Yeah, but now motion data would go from Java to JS, only to go back on the next event loop tick. Sounds like a bad idea, though I have to admit it would simplify merging the data a bit.

**Comment by medape** [comment]
I vote against reverting the changes to Java code, too.

There's a myriad of ways to sidestep this issue, but removing the webserver altogether would keep the connections simpler, avoid an overhead of 12 bytes per state update, and keep each person's configuration in their own device. It would actually make it easier to implement vibration configuration as per #33 (we could combine approaches 2 and 4 by exporting JSON to cookies or localStorage).

Also, I believe we can reduce even further the message length, and every byte helps with that. 32 bytes should be enough for normal layouts. (1 byte per control on layout announcement; a 1-byte prefix, plus 2 bytes per axis and 1/8 byte per button on state update)

The app code will be a little bit more complex, but once the main code is ready, updating the web interface should be business as usual.

**Comment by rmst** [comment]
The webserver + webview was introduced to allow for easy and fast modifications. Having to recompile the app for every change was much more painful and requires the whole Android development suite. 

If you don't want to use a custom WebAppInterface, I suspect there could still be other workarounds that are simpler than downloading files to the phone storage, e.g.

1) Whitelist the address we are using with
https://developer.android.com/reference/android/webkit/WebView.html#setSafeBrowsingWhitelist(java.util.List%3Cjava.lang.String%3E,%20android.webkit.ValueCallback%3Cjava.lang.Boolean%3E)
(Not entirely sure if that would actually work)

2) We could do "adhoc" https, e.g. with flask:
```python
from flask import Flask
app = Flask(__name__)

@app.route("/")
def hello():
    return "Hello World!"

if __name__ == "__main__":
    app.run(ssl_context='adhoc')
```
And then in the WebViewClient (to circumvent invalid certificate messages)
```
@Override
public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
    // ignore ssl error
    if (handler != null){
        handler.proceed();
    } else {
        super.onReceivedSslError(view, null, error);
    }
}
```

**Comment by pzmarzly** [comment]
@rmst 

> Whitelist the address we are using with setSafeBrowsingWhitelist

That probably won't work, unless Safe Browsing here is different from what I expect it is (checking for known scam websites).

Unconditional `proceed()` in `onReceivedSslError` prevents app [from being distributed in Google Play Store](https://stackoverflow.com/questions/5977977/does-the-web-view-on-android-support-ssl)

I'm also against recompiling APK on changes, but if we were to load files from `file://` protocol, we could (but don't have to) use a mechanism other than webserver to transfer the configuration. But I'm not sure what @medape has in mind about that.

**Comment by rmst** [comment]
In the link you posted it said that Google sent a warning email to developers who copy pasted a similar code snippet into their app. This is something I could live with. Transferring and saving everything first would be fine, too.

**Comment by pzmarzly** [comment]
> a warning email

afaik it was a deprecation notice, and [now apps are being rejected outright](https://stackoverflow.com/a/32406156/5108318)

**Comment by medape** [comment]
> I'm also against recompiling APK on changes, but if we were to load files from file:// protocol, we could (but don't have to) use a mechanism other than webserver to transfer the configuration. But I'm not sure what @medape has in mind about that.

File copy-paste. That's why I was asking about this:

> AFAIK apps don't need permissions to write to `/sdcard/Android/data/<appname>/files` (internal storage, called sdcard for legacy reasons) or even `/data/media/<SD-CARD-ID>/Android/data/<appname>/files`. But if I'm wrong, I think app should quit with an error.

I know some paths are readily accessible via MTP and others aren't, but I'm not sure which is which. I try to avoid anything which requires the Android suite, or ADB alone, too. So I'd just have users connect their phones by USB and edit their files from their computer. People who are inclined to edit their configuration will probably know how to edit and send files via USB.

In any case, a tutorial in HTML format would be nice too. Just adding a button that loads the manual instead of a layout would be enough.

**Comment by pzmarzly** [comment]
Closed by https://github.com/rmst/yoke-android/pull/5
